docs/ProcedureExecution.md

```diff
@@ -108,7 +108,45 @@ struct StorageCapReadOutput {
 
 ```
 
-When reading, the parameters are parsed as a list of u256 keys, and for each key a value is returned. When writing, we parse two lists of keys and values to be written. Output in this case are the old values being written over. 
+When reading, the parameters are parsed as a list of u256 keys, and for each key a value is returned. When writing, we parse two lists of keys and values to be written. Output in this case are the old values being written over.
+
+#### Summary of Message Format
+
+The format of the memory message is as follows: 1 byte for the capability type, one byte for the capability index. Each procedure has a list of capabilities, and the capability index is an index into this list. (**TODO**: Why do we need to specify the capability type here? Isn't that contained within the capability?). After that is the start address where we will write, as well as the number of addresses we will write to.
+
+* Byte 0: Capability Type
+* Byte 1: Capability Index
+* Bytes 2 to 34: Number of addresses to be written to (this values is referred to as `n`)
+* Bytes 34 to 34+n\*3`: Storage keys to be written to.
+* Bytes 34+n\*3 to 34+2\*n\*32: Values to be written into each corresponding key.
+
+For example, to write the value `0x34` into the storage key `0x7` the message format will be as follows.
+
+* Byte 0: `0x7`
+* Byte 1: A single byte corresponding to the index of this capability in the list.
+* Bytes 2-33: `0x1`, we are writing to a single address.
+* Bytes 34-65: `0x7`, the single address we are writing to.
+* Bytes 66-97: `0x34`, the single value we are writing into to.
+
+#### Summary of the Capability Format
+
+The write capability is 2 32-byte values. A storage address and an offset/size. The capability allows writing to any value with `address + offset`.
+
+* Byte 0: The capability type (for write it is `0x7`)
+* Bytes 1-32: `address`, the start address.
+* Bytes 33-64: `offset`, the number of addresses (starting with `address` that can be written)
```

Comment:
I would say we should have this indexed to 0, otherwise the capability is empty at 0.

Reply:
So if offset is zero, you cannot even write to the original address. That makes sense to me. Size 0 means nothing. It doesn't limit you in any way, you can still say `0xff..ff` to mean all addresses.

Reply:
I'm not sure if it makes sense to allow a situation where offset is zero, otherwise the capability shouldn't exist.

Reply:
It's a bit of a trade-off, but a "null" capability isn't necessarily a terrible thing, it's not like it would cause an error, it's all very consistent. How than do we describe the second value "additional addresses after the first address"? I don't think using a less obvious definition is worth not having a null capability.

It also make the "write to all" case a bit more obvious, that is `0x00..00,0xff..ff` and avoids the case on the other end where we can write to `0xff..ff` + 1. Not that this is a critical problem either.

For example:

* Write to a single address: `0x7,0x1`, 1 is intuitively one 32-byte value
* Write to all addresses: `0x0,0xff..ff`, all bits off, all bits on.
* Write to `n` addresses: `0x7,n`, `n` is directly the size we can write to.
* Null capability, maybe we do actually want to be able to say this. If we do it is simply 0-filled, which is  a valid capability format.